include/deal.II/base/index_set.h

```diff
@@ -1654,9 +1691,34 @@ IndexSet::add_indices(const ForwardIterator &begin, const ForwardIterator &end)
           ++q;
         }
 
-      add_range(begin_index, end_index);
+      tmp_ranges.emplace_back(begin_index, end_index);
       p = q;
+      if (p != end && *p < end_index)
+        ranges_are_sorted = false;
     }
+
+  if (!ranges_are_sorted)
+    std::sort(tmp_ranges.begin(), tmp_ranges.end());
```

Comment:
Interesting that this works. I guess it uses the lexicographic ordering of the pairs, which for non-overlapping ranges does the right thing...

Reply:
It should also work for overlapping ranges because the lexicographic orderings goes by the first index. We are in an uncompressed state, so the only thing I could imagine is not sorted is if we have two ranges with the same starting index when the one with the higher ending index would be placed afterwards. But that is exactly a case for the merging in `compress()`.